pkg/controllers/management/auth/globalroles/globalrole_handler_test.go

```diff
@@ -0,0 +1,834 @@
+			},
+			globalRole: namespacedRulesGR.DeepCopy(),
+			wantError:  true,
+		},
+		{
+			name: "namespace is not found",
+			stateSetup: func(state grTestState) {
+				state.rListerMock.ListFunc = func(_ string, _ labels.Selector) ([]*rbacv1.Role, error) {
+					return nil, nil
+				}
+
+				nsNotFound := apierrors.NewNotFound(schema.GroupResource{
+					Group:    normanv1.RoleGroupVersionKind.Group,
+					Resource: normanv1.RoleGroupVersionResource.Resource,
+				}, "")
+
+				state.nsCacheMock.EXPECT().Get(gomock.Any()).AnyTimes().Return(activeNamespace, nsNotFound)
+			},
+			globalRole: namespacedRulesGR.DeepCopy(),
+			wantError:  false,
+		},
+		{
+			name: "namespace is nil",
+			stateSetup: func(state grTestState) {
+				state.rListerMock.ListFunc = func(_ string, _ labels.Selector) ([]*rbacv1.Role, error) {
+					return nil, nil
+				}
+				state.nsCacheMock.EXPECT().Get(gomock.Any()).AnyTimes().Return(nil, nil)
+			},
+			globalRole: namespacedRulesGR.DeepCopy(),
+			wantError:  false,
+		},
+		{
+			name: "getting role fails",
+			stateSetup: func(state grTestState) {
+				state.rListerMock.ListFunc = func(_ string, _ labels.Selector) ([]*rbacv1.Role, error) {
+					return nil, nil
+				}
+				state.nsCacheMock.EXPECT().Get(gomock.Any()).AnyTimes().Return(activeNamespace, nil)
+				state.rListerMock.GetFunc = func(_, _ string) (*rbacv1.Role, error) {
+					return nil, fmt.Errorf("error")
+				}
+			},
+			globalRole: namespacedRulesGR.DeepCopy(),
+			wantError:  true,
+		},
+		{
+			name: "creating role fails",
+			stateSetup: func(state grTestState) {
+				state.rListerMock.ListFunc = func(_ string, _ labels.Selector) ([]*rbacv1.Role, error) {
+					return nil, nil
+				}
+				state.nsCacheMock.EXPECT().Get(gomock.Any()).AnyTimes().Return(activeNamespace, nil)
+				state.rListerMock.GetFunc = func(_, _ string) (*rbacv1.Role, error) {
+					return nil, apierrors.NewNotFound(schema.GroupResource{
+						Group:    normanv1.RoleGroupVersionKind.Group,
+						Resource: normanv1.RoleGroupVersionResource.Resource,
+					}, "")
+				}
+				state.rClientMock.CreateFunc = func(role *rbacv1.Role) (*rbacv1.Role, error) {
+					return nil, fmt.Errorf("error")
+				}
+			},
+			globalRole: namespacedRulesGR.DeepCopy(),
+			wantError:  true,
+		},
+		{
+			name: "created role already exists then get fails",
+			stateSetup: func(state grTestState) {
+				state.rListerMock.ListFunc = func(_ string, _ labels.Selector) ([]*rbacv1.Role, error) {
+					return nil, nil
+				}
+				state.nsCacheMock.EXPECT().Get(gomock.Any()).AnyTimes().Return(activeNamespace, nil)
+				state.rListerMock.GetFunc = func(_, _ string) (*rbacv1.Role, error) {
+					if state.counter%2 == 0 {
+						state.counter += 1
+						return nil, apierrors.NewNotFound(schema.GroupResource{
+							Group:    normanv1.RoleGroupVersionKind.Group,
+							Resource: normanv1.RoleGroupVersionResource.Resource,
+						}, "")
+					} else {
+						return nil, fmt.Errorf("error")
+					}
```

Comment:
Nit: I highly recommend not creating tests cases like this which rely on call order to function. This highly couples your test case to the internal implementation of the underlying code, and can create flaky tests which fail when a specific order occurs.

In this case, I would recommend pre-computing the name of each role that might be retrieved, and then return a specific response for the specific roles which are being called.

Reply:
The issue is that the behaviour I'm looking at requires that the `Get` is called twice with the same namespace and role but only fails the second time. I think I could remove the test since it is implementation specific.

Reply:
So in this specific test case I could get rid of the counter since I'm trying to get into the error path. However, I kept a similar pattern for the test at line 224. I've added more comments to explain the reason, and changed the counter to be modified based on when `Create` gets called, not just on consecutive `Get` calls.

I considered removing the test at 224, but I think it's important functionality to test. The difficulty is that the `Get` function gets called with the same input twice, but the expected behaviour changes depending on when it happens. I personally think modifying the counter to indicate before and after `Create` makes it less flaky, but if you feel it shouldn't exist I won't argue much.